File: packages/contracts/v1/spatial-graph.ts
import type { UUID, ISO8601 } from "./common.js";

export interface SpatialNode {
  id: string;
  [key: string]: unknown;
}

export interface SpatialEdge {
  id: string;
  [key: string]: unknown;
}

export interface LevelDescriptor {
  id: string;
  [key: string]: unknown;
}

export interface CirculationGraph {
  [key: string]: unknown;
}

export interface SpatialGraph {
  schema_version: "1.0";
  id: UUID;
  parsed_plan_id: UUID;
  nodes: SpatialNode[];
  edges: SpatialEdge[];
  levels: LevelDescriptor[];
  circulation_graph: CirculationGraph;
  room_adjacency_matrix: number[][];
  scale_meters_per_unit: number;
  coordinate_system: "metric" | "imperial";
  created_at: ISO8601;
}

File: packages/contracts/python/noa_contracts/spatial_graph.py
from __future__ import annotations

from typing import Literal, Any
from pydantic import BaseModel


class SpatialNode(BaseModel):
    id: str
    model_config = {"extra": "allow"}


class SpatialEdge(BaseModel):
    id: str
    model_config = {"extra": "allow"}


class LevelDescriptor(BaseModel):
    id: str
    model_config = {"extra": "allow"}


class SpatialGraph(BaseModel):
    schema_version: Literal["1.0"] = "1.0"
    id: str
    parsed_plan_id: str
    nodes: list[SpatialNode]
    edges: list[SpatialEdge]
    levels: list[LevelDescriptor]
    circulation_graph: dict[str, Any]
    room_adjacency_matrix: list[list[float]]
    scale_meters_per_unit: float = 1.0
    coordinate_system: Literal["metric", "imperial"] = "metric"
    created_at: str
